compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -797,50 +798,62 @@ object Checking {
     tree
 
   /** Check that experimental language imports in `trees`
-   *  are done only in experimental scopes, or in a top-level
-   *  scope with only @experimental definitions.
+   *  are done only in experimental scopes. For for top-level
+   *  experimental imports, all top-level definitions are transformed
+   *  to @experimental definitions.
+   *
    */
-  def checkExperimentalImports(trees: List[Tree])(using Context): Unit =
-
-    def nonExperimentalStat(trees: List[Tree]): Tree = trees match
-      case (_: Import | EmptyTree) :: rest =>
-        nonExperimentalStat(rest)
-      case (tree @ TypeDef(_, impl: Template)) :: rest if tree.symbol.isPackageObject =>
-        nonExperimentalStat(impl.body).orElse(nonExperimentalStat(rest))
-      case (tree: PackageDef) :: rest =>
-        nonExperimentalStat(tree.stats).orElse(nonExperimentalStat(rest))
-      case (tree: MemberDef) :: rest =>
-        if tree.symbol.isExperimental || tree.symbol.is(Synthetic) then
-          nonExperimentalStat(rest)
-        else
-          tree
-      case tree :: rest =>
-        tree
-      case Nil =>
-        EmptyTree
-
-    for case imp @ Import(qual, selectors) <- trees do
+  def checkAndAdaptExperimentalImports(trees: List[Tree])(using Context): Unit =
+    def nonExperimentalTopLevelDefs(pack: Symbol): Iterator[Symbol] =
+      def isNonExperimentalTopLevelDefinition(sym: Symbol) =
+        !sym.isExperimental
+        && sym.source == ctx.compilationUnit.source
+        && !sym.isConstructor // not constructor of package object
+        && !sym.is(Package) && !sym.isPackageObject && !sym.name.endsWith(str.TOPLEVEL_SUFFIX)
+
+      val packageMembers =
+        pack.info.decls
+          .toList.iterator
+          .filter(isNonExperimentalTopLevelDefinition)
+      val packageObjectMembers =
+        pack.info.decls
+          .toList.iterator
+          .filter(sym => sym.isClass && (sym.is(Package) || sym.isPackageObject))
+          .flatMap(nonExperimentalTopLevelDefs)
+
+      packageMembers ++ packageObjectMembers
```

Comment:
Consider combining the two definitions into:
```scala
pack.info.decls.toList.iterator.flatMap: sym =>
  if sym.isClass && (sym.isPackage or sym.isPackageObject) then 
     nonExperimentalTopLevelDefs(sym)
  else if isNonExperimentalTopLevelDefinition(sym) then sym :: Nil
  else Nil
```

Reply:
Updated